ocular/src/keyring.rs

```diff
@@ -0,0 +1,740 @@
+        name: &str,
+        password: &str,
+        derivation_path: Option<&str>,
+        override_if_exists: bool,
+    ) -> Result<PrivateKeyOutput, KeyStoreError>;
+
+    /// Delete key with a given name. If no key exists under name specified an error will be thrown.
+    fn delete_key(&self, name: &str) -> Result<(), KeyStoreError>;
+
+    /// Rename key. If override_if_exists is true any keys with the new key name will be forecfully overriden.
+    fn rename_key(
+        &self,
+        current_name: &str,
+        new_name: &str,
+        override_if_exists: bool,
+    ) -> Result<(), KeyStoreError>;
+
+    /// Get key address in bech32 (aka segwit) format. Will throw an error if the key does not exist.
+    fn get_key_address(&self, name: &str) -> Result<PublicKeyOutput, KeyStoreError>;
+
+    /// Get all key addresses in bech32 (aka segwit) format.
+    fn get_all_keys(&self) -> Result<Vec<PublicKeyOutput>, Box<dyn std::error::Error>>;
+
+    /// Recover key via mnemonic, password, and derivation_path (defaults to cosmos). If override_if_exists is set to true, it will override any existing key with the same name.
+    fn recover_from_mnemonic(
+        &self,
+        name: &str,
+        mnemonic: &str,
+        password: &str,
+        derivation_path: Option<&str>,
+        override_if_exists: bool,
+    ) -> Result<(), KeyStoreError>;
+}
+
+/// Mnemonic and private key in pkcs8 PrivateKeyDocument format
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct PrivateKeyOutput {
+    mnemonic: String,
+    private_key: PrivateKeyDocument,
+}
+
+/// Key name and address in Bech32 (aka segwit) format
+// TODO: Include actual public key here if ever needed
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct PublicKeyOutput {
+    name: String,
+    address: String,
+}
+
+// --- Base Key Ring ---
+/// Base Keyring that needs to be initialized before being used. Initialization parameters vary depending on type of key store being used.
+#[allow(dead_code)]
+pub struct Keyring {
+    backend: KeyRingType,
+    pub key_store: Box<dyn KeyStore>,
+}
+
+// Keyring constructors
+impl Keyring {
+    /// Create new instance of FsKeyStore Keyring
+    /// Will create store at current working dir if None is provided
+    pub fn new_file_store(key_path: Option<&str>) -> Self {
+        let path: String;
+
+        if let Some(key_path) = key_path {
+            path = key_path.to_string();
+        } else {
+            path = env::current_dir()
+                .unwrap()
+                .into_os_string()
+                .into_string()
+                .expect("Could not get current working dir.");
+        }
+
+        dbg!(format!("Attempting to use path {}", path));
+
+        let mut key_store = FileKeyStore {
+            key_path: path.clone(),
+            key_store: None,
+        };
+
+        key_store
+            .create_key_store()
+            .unwrap_or_else(|_| panic!("Could not create file key store for path: {}", path));
```

Comment:
That's a good point. I guess the question is how recoverable this error can be. If we fail to create the keystore would we prefer to be able to try to recover via an error, or should this be viewed as something we need to exit immediately for? (don't think there's necessarily a right answer to this one, more so the flexibility/pattern we want)

Reply:
Since this is a library that will be publicly consumed I think that's a decision best left to users. If they want to panic on keystore error they can do that but I think it seems like surprising behavior by default.